### PR TITLE
Fill Mode added

### PR DESCRIPTION
**Pull request by HannahRKent**

### CODE REVIEW

File: src/main/java/de/orchound/gameoflife/view/Painter.java [hunk 1]
```diff
@@ -1,19 +1,30 @@
...
+	private boolean fillMode = false;
 	private boolean paintMode = false;
 
 	public Painter(Game game) {
 		this.game = game;
 	}
 
-	public void paintOrEraseCell(Vector2ic cell) {
+	public void paintAction(Vector2ic cell) {
+		if (fillMode) {
```
Comment (kBlaszczyk) [inline src/main/java/de/orchound/gameoflife/view/Painter.java]: The painting algorithms shouldn't be implemented in the same class.
Create a `Painter` interface and then a `TogglePainter` for the original implementation and a `FillPainter` for filling in cells.
The painters should be organized in the `Applet` and selected if an appropriate button is being clicked.

**Review by kBlaszczyk — CHANGES_REQUESTED** [review]
I think the application crashes when the filling process reaches the board limits.
Try to fill after clearing out everything.

File: src/main/java/de/orchound/gameoflife/GameOfLifeApplet.java [hunk 2]
```diff
@@ -24,6 +27,9 @@
 
 	private final Vector2i windowSize = new Vector2i(1280, 720);
 
+	// private PainterPicker painterPicker;
```
Comment (kBlaszczyk) [inline src/main/java/de/orchound/gameoflife/GameOfLifeApplet.java]: remove line
```suggestion
```

File: src/main/java/de/orchound/gameoflife/GameOfLifeApplet.java [hunk 3]
```diff
@@ -143,7 +153,19 @@ public void keyPressed() {
...
 		case 's' -> game.save();
+		case 'f' -> setFillPainter();
+		case 'p' -> setTogglePainter();
 		case BACKSPACE -> game.clear();
 		}
 	}
+
+	private void setFillPainter() {
+		boardView.setPainter(fillPainter);
+		System.out.println("fill mode");
```
Comment (kBlaszczyk) [inline src/main/java/de/orchound/gameoflife/GameOfLifeApplet.java]: remove the debug logs
```suggestion
```

File: src/main/java/de/orchound/gameoflife/GameOfLifeApplet.java [hunk 4]
```diff
@@ -143,7 +153,19 @@ public void keyPressed() {
...
+	}
+
+	private void setTogglePainter() {
+		boardView.setPainter(togglePainter);
+		System.out.println("toggle mode");
```
Comment (kBlaszczyk) [inline src/main/java/de/orchound/gameoflife/GameOfLifeApplet.java]: remove the debug logs
```suggestion
```

File: src/main/java/de/orchound/gameoflife/model/Game.java [hunk 5]
```diff
@@ -98,11 +98,22 @@ public void setCell(Vector2ic cell, boolean resurrect) {
...
 	}
 
+	public void setCellSilently(Vector2ic cell, boolean resurrect) {
+		if (cellInBoardRange(cell) && getCellStatus(cell) != resurrect) {
+			if (resurrect) {
+				board.resurrectCell(cell.y(), cell.x());
+			} else {
+				board.killCell(cell.y(), cell.x());
+			}
+			notifyBoardDataObservers();
```
Comment (kBlaszczyk) [inline src/main/java/de/orchound/gameoflife/model/Game.java]: The point was to not notify the observers when setting a cell silently, right?
```suggestion
```

File: src/main/java/de/orchound/gameoflife/model/Game.java [hunk 6]
```diff
@@ -98,11 +98,22 @@ public void setCell(Vector2ic cell, boolean resurrect) {
...
+		}
+	}
+
 	public boolean getCellStatus(Vector2ic cell) {
 		return cellInBoardRange(cell) && board.getCellStatus(cell.y(), cell.x());
 	}
 
-	private boolean cellInBoardRange(Vector2ic cell) {
+	public boolean cellInBoardRange(Vector2ic cell) {
```
Comment (kBlaszczyk) [inline src/main/java/de/orchound/gameoflife/model/Game.java:115]: The board range validation is done internally when setting a cell.
Make this private again and remove the method call from the `FillPainter`.

File: src/main/java/de/orchound/gameoflife/painting/FillPainter.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+	public FillPainter(Game game) {
+		this.game = game;
+	}
+
+	private void fillCells(Vector2ic cell) {
+		if (game.getCellStatus(cell) != paintMode && game.cellInBoardRange(cell)) {
+			game.setCellSilently(cell, paintMode);
+			fillCells(new Vector2i(cell.x(), cell.y()+1));
+			fillCells(new Vector2i(cell.x(), cell.y()-1));
```
Comment (kBlaszczyk) [inline src/main/java/de/orchound/gameoflife/painting/FillPainter.java]: Fix operator spaces.
```suggestion
			fillCells(new Vector2i(cell.x(), cell.y() + 1));
			fillCells(new Vector2i(cell.x(), cell.y() - 1));
```

File: src/main/java/de/orchound/gameoflife/view/Painter.java [hunk 8]
```diff
@@ -1,19 +1,30 @@
 package de.orchound.gameoflife.view;
 
 import de.orchound.gameoflife.model.Game;
+import org.joml.Vector2i;
 import org.joml.Vector2ic;
 
 public class Painter {
```
Comment (kBlaszczyk) [inline src/main/java/de/orchound/gameoflife/view/Painter.java]: You should be able to remove the `Painter` class now.

File: src/test/java/de/orchound/gameoflife/painting/FillPainterTest.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+	void setUp() {
+		game.clear();
+	}
+
+	@Test
+	void testPaint() {
+		game.clear();
+
+		for (int i = 0; i < 3; i ++) {
+			game.toggleCell(new Vector2i(i, 2));
```
Comment (kBlaszczyk) [inline src/test/java/de/orchound/gameoflife/painting/FillPainterTest.java]: Using `setCell()` here and below is more explicit.

File: src/test/java/de/orchound/gameoflife/painting/TogglePainterTest.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+	Game game = new Game(6, 6);
+	TogglePainter painter = new TogglePainter(game);
+	@BeforeEach
+	void setUp() {
+		game.clear();
+	}
+
+	@Test
+	void testPaint() {
+		game.clear();
```
Comment (kBlaszczyk) [inline src/test/java/de/orchound/gameoflife/painting/TogglePainterTest.java]: Redundant clear call.
```suggestion
```

File: src/test/java/de/orchound/gameoflife/painting/TogglePainterTest.java [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+		Vector2ic cell = new Vector2i(2, 2);
+		painter.paint(cell);
+
+		assertTrue(game.getCellStatus(new Vector2i(2, 2)));
+
+
+		for (int i = 0; i <= game.getSize().x(); i++) {
```
Comment (kBlaszczyk) [inline src/test/java/de/orchound/gameoflife/painting/TogglePainterTest.java]: This test only verifies the setting of a cell, not the actual toggle logic.
Maybe just toggle an entire random board and then verify that every cell changed status.

File: src/test/java/de/orchound/gameoflife/view/BoardViewTest.java [hunk 12]
```diff
@@ -29,13 +31,15 @@
 	@Spy
 	private final Game game = new Game(10, 5);
 
+	private final Painter painter = new TogglePainter(game);
```
Comment (kBlaszczyk) [inline src/test/java/de/orchound/gameoflife/view/BoardViewTest.java]: No need for this member in this test.

File: src/test/java/de/orchound/gameoflife/view/PainterTest.java [hunk 13]
```diff
@@ -0,0 +1,77 @@
+//package de.orchound.gameoflife.view;
```
Comment (kBlaszczyk) [inline src/test/java/de/orchound/gameoflife/view/PainterTest.java]: File can be removed

**Review by kBlaszczyk — CHANGES_REQUESTED** [review]

File: src/main/java/de/orchound/gameoflife/model/Game.java (hunk 6 above)
Comment (HannahRKent) [reply src/main/java/de/orchound/gameoflife/model/Game.java:115]: Removing the method call causes my tests to fail and the program to crash

File: src/test/java/de/orchound/gameoflife/painting/TogglePainterTest.java (hunk 11 above)
Comment (HannahRKent) [reply src/test/java/de/orchound/gameoflife/painting/TogglePainterTest.java]: Wasn't sure what you meant, I don't think the togglepainter actually uses toggleCell.  Maybe we have bad naming?  Anyway I made the test better

File: src/test/java/de/orchound/gameoflife/view/BoardViewTest.java [hunk 14]
```diff
@@ -35,7 +36,7 @@
 	public void setUp() {
 		when(sketch.createImage(10, 5, PConstants.RGB))
 			.thenReturn(new PImage(10, 5));
-		boardView = new BoardView(game, sketch, boardRenderer);
+		boardView = new BoardView(game, sketch, boardRenderer, new TogglePainter(game));
```
Comment (kBlaszczyk) [inline src/test/java/de/orchound/gameoflife/view/BoardViewTest.java]: Using a `Painter` mock here would be nice.

File: src/test/java/de/orchound/gameoflife/painting/FillPainterTest.java [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+	FillPainter fillPainter = new FillPainter(game);
+
+	@BeforeEach
+	void setUp() {
+		game.clear();
+	}
+
+	@Test
+	void testPaint() {
+		game.clear();
```
Comment (kBlaszczyk) [inline src/test/java/de/orchound/gameoflife/painting/FillPainterTest.java]: Remove redundant `clear()` call.
```suggestion
```

File: src/test/java/de/orchound/gameoflife/painting/TogglePainterTest.java [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+	TogglePainter painter = new TogglePainter(game);
+	@BeforeEach
+	void setUp() {
+		game.clear();
+		game.setCell(new Vector2i(2, 1), true);
+		game.setCell(new Vector2i(2, 3), true);
+	}
+
+	@Test
+	void testPaint() {
```
Comment (kBlaszczyk) [inline src/test/java/de/orchound/gameoflife/painting/TogglePainterTest.java]: This test could be better. Try implementing the following scenario.
```
oxo    xxx
xox => ooo
^
```
The painting is started on the first cell for both lines.
(each line is representing a single painting process, consider creating two test methods)

File: src/test/java/de/orchound/gameoflife/painting/FillPainterTest.java [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+	Game game = new Game(6, 6);
+	FillPainter fillPainter = new FillPainter(game);
+
+	@BeforeEach
+	void setUp() {
+		game.clear();
+	}
+
+	@Test
+	void testPaint() {
```
Comment (kBlaszczyk) [inline src/test/java/de/orchound/gameoflife/painting/FillPainterTest.java]: Please implement two test methods `testFillBoard()` and `testFillSquare()`.
The former simply fill paints any cell of an empty board and verifies that the entire board is set.
The latter could paint a 4x4 square on the 6x6 board and then fill paint that square. (Then verifying that only the square and its' content was set)

**Review by kBlaszczyk — CHANGES_REQUESTED** [review]

File: src/test/java/de/orchound/gameoflife/painting/FillPainterTest.java [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+		for (int i = 2; i < 5; i++) {
+			game.setCell(new Vector2i(1, i), true);
+		}
+
+		for (int i = 2; i < 5; i++) {
+			game.setCell(new Vector2i(4, i), true);
+		}
+
+		game.setCell(new Vector2i(2, 4), true);
+		game.setCell(new Vector2i(3, 4), true);
```
Comment (kBlaszczyk) [inline src/test/java/de/orchound/gameoflife/painting/FillPainterTest.java]: Simplify the square initialization like this:
```suggestion
for (int i = 1; i < 5; i++) {
	game.setCell(new Vector2i(i, 1), true);
	game.setCell(new Vector2i(1, i), true);
	game.setCell(new Vector2i(4, i), true);
	game.setCell(new Vector2i(i, 4), true);
}
```

File: src/test/java/de/orchound/gameoflife/painting/FillPainterTest.java [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+		for (int i = 0; i < game.getSize().x(); i++) {
+			for (int j = 5; j < game.getSize().y(); j++) {
+				assertFalse(game.getCellStatus(new Vector2i(i, j)));
+			}
+		}
+
+		for (int i = 1; i < 4; i++) {
+			assertFalse(game.getCellStatus(new Vector2i(0, i)));
+			assertFalse(game.getCellStatus(new Vector2i(5, i)));
+		}
```
Comment (kBlaszczyk) [inline src/test/java/de/orchound/gameoflife/painting/FillPainterTest.java]: Verifying the immediate surroundings of the square can also be simplified like this:
```suggestion
for (int i = 0; i < 6; i++) {
	assertFalse(game.getCellStatus(new Vector2i(0, i)));
	assertFalse(game.getCellStatus(new Vector2i(i, 0)));
	assertFalse(game.getCellStatus(new Vector2i(5, i)));
	assertFalse(game.getCellStatus(new Vector2i(i, 5)));
}
```

**Review by kBlaszczyk — CHANGES_REQUESTED** [review]

**Review by kBlaszczyk — APPROVED** [review]

**Pull request merged**